### PR TITLE
Improve Promise type definition

### LINKED ISSUE #4903: somePromise.catch or .then(null, rejectFn) should preserve type of original promise but doesn't
The fundamental point is that the `catch` of a promise should NOT change the type of that promise unless it returns something different.  Returning something different is relatively rare because the activation of the `catch` should be rare. 

Because `catch` is sugar over `then(null, rejectFn)`, that form has the same issue.

Example 1: foo returns Promise&lt;Hero>
```
	foo(){
		return Promise.resolve(new Hero('foo'));
	}
```

Example 2: foo returns a Promise&lt;{}> !  This is AWFUL
```
	foo(){
		return Promise.resolve(new Hero('foo'))
                  // Pick any of these to see the problem
		  .catch()
		  .catch(e => {})
		  .then(null, e => {})
	}
```
Example 3: foo returns a Promise&lt;{}> !  This is AWFUL TOO
```
	foo(){
		return Promise.resolve(new Hero('foo'))
                  // Pick either of these to see the problem
		  .catch(e => Promise.reject(e))
		  .then(null, e => Promise.reject(e))
	}
```

Example 4: foo returns a Promise&lt;Hero>  ... but this is baroque
```
	foo(){
		return Promise.resolve(new Hero('foo'))
		  .then(_ => _, e => Promise.reject(e));
	}
```
The problem is in the typings file I think.  I contend that that `catch` or `then(null, rejectFn)`  should be typed to return a Promise&lt;Hero>.

### PR DESCRIPTION
**Pull request by rbuckton**
This change modifies the overloads for `then` and `catch` of `Promise<T>`, and the overloads for `then` of `PromiseLike<T>` to provide better type inference.

Fixes #4903

### CODE REVIEW

**Comment by rbuckton** [comment]
@sandersn, @vladima: care to take a look?

**Comment by vladima** [comment]
:+1:

File: src/lib/es2015.promise.d.ts [hunk 1]
```diff
@@ -3,47 +3,76 @@
  */
 interface Promise<T> {
     /**
+     * Creates a new Promise with the same internal state of this Promise.
+     * @returns A Promise.
+     */
+    then(): Promise<T>;
```
Comment (sandersn) [inline src/lib/es2015.promise.d.ts]: I thought overloads should be arranged longest-to-shortest to make sure that the most restrictive overloads matches first. Why change that here?

**Comment by rbuckton** [comment]
The change in 9d4219a fixes [an existing issue](https://github.com/Microsoft/TypeScript/issues/10524#issuecomment-242586264) with assignability between `Promise<T>` and `PromiseLike<T>`. Unfortunately, this means that some cases in this change that should result in `Promise<never>` retain the `T` of the antecedent promise. This is a result of how we erase uninstantiated generic type parameters to `any` and how `any` compares to `never`. A more comprehensive solution to address this is out of scope of this PR, as this still provides an improvement over our existing type definition for `Promise<T>`.

File: src/lib/es2015.promise.d.ts [hunk 2]
```diff
@@ -8,42 +8,45 @@ interface Promise<T> {
      * @param onrejected The callback to execute when the Promise is rejected.
      * @returns A Promise for the completion of which ever callback is executed.
      */
-    then<TResult1, TResult2>(onfulfilled: (value: T) => TResult1 | PromiseLike<TResult1>, onrejected: (reason: any) => TResult2 | PromiseLike<TResult2>): Promise<TResult1 | TResult2>;
+    then(onfulfilled?: ((value: T) => T | PromiseLike<T>) | undefined | null, onrejected?: ((reason: any) => T | PromiseLike<T>) | undefined | null): Promise<T>;
```
Comment (rbuckton) [inline src/lib/es2015.promise.d.ts]: This overload is meant to replace `then(): Promise<T>` as that overload was one of the causes of #10524.

**Comment by rbuckton** [comment]
@sandersn, @vladima can you take another look at these typings? The order of the overloads is intentional and very specific to get the best typings with respect to `never` and how `undefined | null` are handled depending on the value of `strictNullChecks`.

**Comment by mhegazy** [comment]
👍

**Comment by rbuckton** [comment]
@mhegazy, do you want to take this change for release-2.0 as well?

**Comment by mhegazy** [comment]
no. jsut 2.1 please.

**Pull request merged**